File: backend/app/schemas/danger_zone.py
from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field, model_validator


class DangerZoneQuery(BaseModel):
    west: float = Field(ge=-180, le=180)
    south: float = Field(ge=-90, le=90)
    east: float = Field(ge=-180, le=180)
    north: float = Field(ge=-90, le=90)

    @model_validator(mode="after")
    def validate_viewport(self):
        if self.west >= self.east:
            raise ValueError("west must be less than east")

        if self.south >= self.north:
            raise ValueError("south must be less than north")

        return self


class DangerZoneCellResponse(BaseModel):
    cell_id: str
    grid_x: int
    grid_y: int
    latitude: float
    longitude: float

    south: float
    west: float
    north: float
    east: float

    incident_count: int
    incident_score: float = Field(ge=0, le=1)
    coverage_ratio: float = Field(ge=0, le=1)
    coverage_sparsity: float = Field(ge=0, le=1)
    danger_score: float = Field(ge=0, le=1)


class DangerZoneData(BaseModel):
    neighbourhood_id: UUID
    window_start: date | None
    window_end: date | None
    calculated_at: datetime | None
    cell_size_metres: int = 100
    min_score: float = Field(ge=0, le=1)
    max_score: float = Field(ge=0, le=1)
    cells: list[DangerZoneCellResponse]


class DangerZoneResponse(BaseModel):
    status: int
    message: str | None = None
    data: DangerZoneData

File: frontend/src/lib/validators/danger-zone.ts
import { z } from "zod";

const DateOnlySchema = z
  .string()
  .regex(/^\d{4}-\d{2}-\d{2}$/);

const ScoreSchema = z
  .number()
  .finite()
  .min(0)
  .max(1);

export const DangerZoneCellSchema = z.object({
  cell_id: z.string().min(1),
  grid_x: z.number().int(),
  grid_y: z.number().int(),

  latitude: z.number().finite().min(-90).max(90),
  longitude: z.number().finite().min(-180).max(180),

  south: z.number().finite().min(-90).max(90),
  west: z.number().finite().min(-180).max(180),
  north: z.number().finite().min(-90).max(90),
  east: z.number().finite().min(-180).max(180),

  incident_count: z
    .number()
    .int()
    .nonnegative(),

  incident_score: ScoreSchema,
  coverage_ratio: ScoreSchema,
  coverage_sparsity: ScoreSchema,
  danger_score: ScoreSchema,
});

export const DangerZoneDataSchema = z.object({
  neighbourhood_id: z.string().uuid(),
  window_start: DateOnlySchema.nullable(),
  window_end: DateOnlySchema.nullable(),

  calculated_at: z
    .string()
    .datetime({
      offset: true,
    })
    .nullable(),

  cell_size_metres: z
    .number()
    .int()
    .positive(),

  min_score: ScoreSchema,
  max_score: ScoreSchema,

  cells: z.array(DangerZoneCellSchema),
});

export const DangerZoneResponseSchema =
  z.object({
    status: z.number().int(),
    message: z.string().nullable().optional(),
    data: DangerZoneDataSchema,
  });

export type DangerZoneCell = z.infer<
  typeof DangerZoneCellSchema
>;

export type DangerZoneData = z.infer<
  typeof DangerZoneDataSchema
>;

export interface DangerZoneViewport {
  west: number;
  south: number;
  east: number;
  north: number;
}